### ISSUE TITLE
Simple explanation for Local Degree Score formula for edges

### ISSUE THREAD

**Issue by j-adamczyk** [networkit/networkit#1028]
Local Degree Score has been created in the context of graph sparsification, as described [in the original article](https://arxiv.org/pdf/1505.00564.pdf). I understand the intuition behind the method, and explanation seems reasonable:

For each node v ∈ V , we include the edges to the top floor(deg(v)^α) neighbors, sorted by degree in descending order. Similar to Local Similarity we use again 1 − α for the minimum parameter α such that an edge is still contained in the sparsified graph as edge score. The goal of this approach is to keep those edges in the sparsified graph that lead to nodes with high degree, i.e. the hubs that are crucial for a complex network’s topology.

However, looking at [the code](https://github.com/networkit/networkit/blob/master/networkit/cpp/sparsification/LocalDegreeScore.cpp), I do not understand this any more. What exactly is the formula for the edge score? Not for sparsification or top k edges, but the raw score for the edge. I can extract them with `.scores()` method, but what exactly is the formula for a single edge?

As far as I can tell, the code for each edge calculates the score for both nodes at its ends and takes the higher score. For each node, its neighbors are sorted ascending by their degree and for each neighboring node we calculate:
- `rank` - number of nodes with same or lower degree as the neighbor currently visited; this seems affected by sorting of neighbors with the same degree, though, which is weird
- `d` - degree of the node itself
And the score for the edge is `1 - log(rank) / log(d)`. I am quite confused about why `rank` looks like this. And, as far as I can tell, the alpha from the original formula is not used, which kind of makes sense, since there is no application to sparsification at this point.

**Comment by fabratu** [comment]
The original paper doesn't include an explanation of how the edge scores (e.g. α) are computed. In the extended [journal version](https://link.springer.com/article/10.1007/s13278-016-0332-2) you can see the rules for what is called the "Local Filtering Score" in Definition 4. 

Basically the edge scores for an edge (u,v) is:

- 1, if d(u)=1
- 1 - log(rank(u,v))/log(d(u))

As you already pointed out, each edge is visited twice. And the final result is the max of both visits. 

> For each node, its neighbors are sorted ascending by their degree ...

One minor correction to your observations. The neighbors are sorted backwards due to the "less than" operator overload in the `AttributizedEdge` being inverted.

**Comment by j-adamczyk** [comment]
@fabratu so neighbors are sorted descending by degree? That would make more sense from what I understood from the paper. So what does a rank mean precisely in this context? Number of neighbors with higher degree? In other words, how to define `r_c` for Local Degree Score?

**Comment by fabratu** [comment]
> @fabratu so neighbors are sorted descending by degree? That would make more sense from what I understood from the paper. So what does a rank mean precisely in this context? Number of neighbors with higher degree? In other words, how to define `r_c` for Local Degree Score?

Yes, as a result they are sorted in descending order.

Concerning the rank, the journal paper has the following defintion:

```
Definition 3 (Neighborhood rank): Given a graph G=(V,E) and 
an edge score c: E -> R, the neighborhood rank rank(u,v) 
is the position of v in the sorted list of neighbors of u, i.e. 
rank(u,v) =|{x in N(u) : c(u,x) < c(u,v)}| + 1.

Note that when two edges that are incident to u have the 
same edge score, they also have the same rank.
```

If I am not mistaken, It might be the case that the definition is not 100% correct. According to the Pseudocode (and implementation) the rank formula should be: `rank(u,v) =|{x in N(u) : c(u,v) < c(u,x)}| + 1` (as the rank is lower for higher scores). Maybe one the authors like @michitux can go more into detail here.

**Comment by j-adamczyk** [comment]
@fabratu I think that it actually makes sense as it is - if you sort descending, then `rank(u,v) =|{x in N(u) : c(u,x) < c(u,v)}| + 1` means that the rank is the number of neighbors with the higher degree. And since the lower the rank, the higher the score, this will results in high score for nodes that have as little neighbors with higher degree as possible. So hubs and nodes with degree 1 will get highest scores.

**Comment by fabratu** [comment]
> if you sort descending, then `rank(u,v) =|{x in N(u) : c(u,x) < c(u,v)}| + 1` means that the rank is the number of neighbors with the higher degree.

Yeah, I think we can agree on the meaning (with the minor addition of a `+1`). Never mind, maybe I just misread the formula or misinterpret what the score-input `c(u,v)` is here.

- If `c(u,v)` for rank computation is the final score, then the score is lower if a rank is higher. This means the degree of v is lower. If we count all neighbors with lower degree, a neighbor with high degree would get a higher rank number (which doesn't make sense).
- If `c(u,v)` for rank computation is the degree itself, then we would again count all nodes with a lower degree - resulting in a higher rank number.
